---
title: "Zika DESeq Demo"
author: "Generated by Bryce Watson for Van Doorslaer Lab"
date: "1/18/2021"
output:
  pdf_document:
    latex_engine: xelatex
    highlight: pygments
bibliography: zika.bib
urlcolor: blue
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, cache=TRUE,
    fig.align='center',
    dev = 'pdf')
```

\hypersetup{linkcolor=black}
\tableofcontents
\pagebreak

## Setup and Data Import
Instruction and rationale for how to conduct RNA-seq using the DESeq2 R package is outlined in the Bioconductor vignette: **[Analyzing RNA-seq data with DESeq2](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts)**.
Script developed with guidance of **[Simon Cockell](https://github.com/sjcockell/lockdown-learning)**.


First we load the library consisting of BiocManager,DESeq2, Biobase, ggplot2, ggrepel, apeglm, pheatmap, genefilter,plotly, tibble, and rmarkdown. Then we perform initial data import of raw count files from a csv.

The dataset for this RNA Seq experiment can be found on the **[NCBI GEO Database](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125554)**

```{r setup, include=FALSE}
library(BiocManager)
library(DESeq2)
library(Biobase)
library(biomaRt)
library(ggplot2)
library(ggrepel)
library(apeglm)
library(pheatmap)
library(genefilter)
library(plotly)
library(tibble)
library(rmarkdown)
library(tinytex)
library(knitr)

#Import count data
setwd("~/R/R_Projects/DESeq2")
zika_counts <- file.path("./Zika_DESeq/GSE125554_zika_counts.csv")
cts <- read.csv(zika_counts, row.names=1, stringsAsFactors=FALSE)
```

## Make DESeq Dataset

Once un-normalized read counts are loaded from a .csv some simple organization is required before performing DESeq operations. In order to make a DESeq Data Set (dds), a .txt or .csv meta file specifying the experimental conditions and samples needs to be built and loaded into the 'col_data' vector.

```{r, include=FALSE}
# DESeq Parameters
countdata <- as.matrix(cts)
zika_meta <- file.path("./Zika_DESeq/zika_metadata.csv")
col_data <- read.csv(zika_meta)
```

Once that is done we can build the basic dds object with the function described below. Then using the results() function we can generate a results table with log2 fold changes, p values and adjusted p values.

```{r, echo=TRUE}
dds = DESeqDataSetFromMatrix(countData = cts,
                             colData = col_data,
                             design = ~ Condition)

#dds prefiltering
ddsf <- dds[ rowSums(counts(dds)) > 1, ] 

#DeSeq Dataset
dds <- DESeq(ddsf)
res <- results(dds)
res_df <- as.data.frame(res)
filter_df <- res_df[complete.cases(res_df),] # Filters out incomplete rows.
```

```{r, echo=FALSE}
kable(filter_df[1:5, ], caption="Header of Results Dataframe")
```

## Normalization and Data Merging

The next step is to normalize data with the VarianceStabilizingTransformation() to produce a *DESeqTransform* object. This object will be mainly used for variance measurements going forward such as a PCA plot, MA plot and heatmap.

```{r, echo=TRUE}
#Normalization
vst = varianceStabilizingTransformation(dds)
vsd <- vst(dds, blind = FALSE)
```

```{r, include=FALSE}
# Merge with normalized count data
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
```

## Boxplot

An easy start to get an overview of the data is with a  simple boxplot of the counts frequency.

```{r, echo = FALSE}
boxplot(assay(vst))
```

## P-Value Histogram

A histogram of the distribution of P-values in the DESeq results table allows us to see a simplified pattern of p-value spread.

```{r, echo = FALSE}
hist(res$pvalue, breaks=40, col='gray', main=paste('Histogram of' , 'P-values'), xlab = 'P-values')
```

## PCA Plot

The first overview we will use to visualize distances between samples  is a principle component analysis (PCA). The plotPCA() function comes with the DESeq2 package and is built on ggplot2. So, it is also compatible with ggplot-adjacent packages such as ggrepel. Here we use the labeling functions of these packages to differentiate between groups.

```{r, echo=TRUE}
plotPCA(vst, intgroup='Condition') +
  geom_text_repel(aes(label=name)) +
  ggtitle("Principle Component Analysis") +
  theme(legend.position="none", plot.title = element_text(size = rel(1.5), hjust = 0.5))
```

## MA Plot

The next step is to get an overview of the experiment with plotMA(). MA plots represent each gene as a point on a graph. The X-axis plots the mean of the gene's expression across all samples. The Y-axis plots the average of counts normalized by size factor or the log2 fold change. The default alpha threshold for adjusted p-values is 0.1, adjusted here to match the 0.05 padj values in later plots.

```{r, include = FALSE}
#Shrink Log-fold change
# 'coef' needs to be found manually with command: 'resultsNames(dds)'
resultsNames(dds)
res <- lfcShrink(dds, coef="Condition_infected_vs_control", type="apeglm")
```

```{r, echo=TRUE}
plotMA(res, alpha=0.05, main='MA Plot', ylim=c(-14,6),
       colNonSig = "gray",
       colSig = "orangered",
       colLine = "grey60")
```

## Gene Clustering Heatmap

For the next two plots we will need to convert Ensembl IDs to more readable gene IDs. Here the biomaRt package is used to convert to HUGO gene nomenclature (hgnc). Other R packages such as the organism annotation package can accomplish similar conversions. 

```{r, echo=TRUE}
#Genes of Interest Annotation
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 25)
mat  <- assay(vsd)[ topVarGenes, ]
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
mat  <- mat - rowMeans(mat)
gns <- getBM(c("hgnc_symbol", "ensembl_gene_id"), "ensembl_gene_id", row.names(mat), mart=mart, useCache=FALSE)
row.names(mat)[match(gns[,2], row.names(mat))] <- gns[,1]
```

Using the annotated gene IDs generated above we can now build a heatmap very simply with the pheatmap package. The 25 most variable genes are highlighted, but this is easily changed my modifying the 'topVarGenes' variable and running the Ens ID to hgnc conversion again.

```{r, echo=TRUE}
#Make Heatmap
anno <- as.data.frame(colData(vsd)[, c("Sample","Condition")])
pheatmap(mat, annotation_col = anno)
```

## Volcano Plot

The final plot to get an overview of the data is a Volcano Plot. The Log2Fold change is on the X-axis. This shows how much a gene has changed. The djusted P-value of each gene is on the Y-axis, showing how significant that change is. This gives a good spread of change in gene expression for our dataset along with the chosen measure of significance (p-adj in this case).

```{r, include=FALSE}
#Filtering
filter_df_padj <- filter_df[filter_df$padj < 0.05,] #Filters by p-value < 0.05
filter_df$test = filter_df$padj < 0.05 & abs(filter_df$log2FoldChange) > 1
#Order by padj value
filter_df=filter_df[order(filter_df$padj),]
filter_df_padj=filter_df_padj[order(filter_df_padj$padj),]

#convert ENS IDs to hgnc symbol for filter_df_padj
topSigGenes <- filter_df_padj[1:20,] # 20 lowest adjusted p-values.

gns_padj <- getBM(c("hgnc_symbol", "ensembl_gene_id"), "ensembl_gene_id", row.names(topSigGenes), mart=mart, useCache=FALSE)
row.names(topSigGenes)[match(gns_padj[,2], row.names(topSigGenes))] <- gns_padj[,1]
topSigGenes = rownames_to_column(topSigGenes, var='Gene')
```

```{r, include=TRUE}
#Volcanoplot
ggplot(filter_df, aes(x=log2FoldChange, y=-log10(padj))) +
  geom_point(aes(color=test), size=1.5, alpha=0.4) +
  scale_color_manual(values=c('violetred', 'gray', 'orangered')) +
  xlim(-15, 20) +
  ggtitle('Volcano Plot') +
  labs(y=expression('-Log'[10]*' P'[adj]), x=expression('Log'[2]*' fold change')) +
  geom_text_repel(data=topSigGenes, force=5,aes(x = log2FoldChange, y = -log10(padj),label=Gene))+
  geom_point(data=topSigGenes,aes(x = log2FoldChange, y = -log10(padj), color='black',alpha=0.4))+
  theme_minimal() +
  theme(legend.position="none", plot.title = element_text(size = rel(1.5), hjust = 0.5))
```

```{r, include=FALSE}
knitr::opts_chunk$set(warning=TRUE)
```

## Conclusion

RNA-Seq analysis shows high variability and significance in similar gene groups. Substantial overlap in findings with the original research article shows this exploratory analysis to be a success. As reported in [@Aguiareaay] collagen-encoding genes and genes which code for extra-cellular matrix proteins are among the most variable. This supports the theory that some Zika congenital abnormalities are associated with increased permeability of the blood-brain barrier. This may heighten risk of ischemic stroke, intracranial bleeding, and allow for permeation of Zika virus into developing neonate cells.

## References

---
nocite: '@*'
---
